js/src/hooks/useSaveShippingTimes.js

```diff
@@ -0,0 +1,99 @@
+/**
+ * External dependencies
+ */
+import { useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { useAppDispatch } from '.~/data';
+import useShippingTimes from './useShippingTimes';
+import getDeletedShippingTimes from '.~/utils/getDeletedShippingTimes';
+import getDifferentShippingTimes from '.~/utils/getDifferentShippingTimes';
+
+/**
+ * @typedef { import(".~/data/actions").ShippingTime } ShippingTime
+ * @typedef { import(".~/data/actions").AggregatedShippingTime } AggregatedShippingTime
+ */
+
+/**
+ * Get the country codes of the deleted shipping times.
+ *
+ * Deleted shipping times are those that exist in `oldShippingTimes` but not in `newShippingTimes`.
+ *
+ * @param {Array<ShippingTime>} newShippingTimes New shipping times.
+ * @param {Array<ShippingTime>} oldShippingTimes Old shipping times.
+ * @return {Array<string>} Array of country codes.
+ */
+const getDeletedCountryCodes = ( newShippingTimes, oldShippingTimes ) => {
+	const deletedShippingTimes = getDeletedShippingTimes(
+		newShippingTimes,
+		oldShippingTimes
+	);
+
+	return deletedShippingTimes.map(
+		( shippingTime ) => shippingTime.countryCode
+	);
+};
+
+/**
+ * Get aggregated shipping time groups from shipping times.
+ *
+ * @param {Array<ShippingTime>} shippingTimes Array of shipping time.
+ * @return {Array<AggregatedShippingTime>} Array of shipping time group.
+ */
+const getShippingTimesGroups = ( shippingTimes ) => {
+	const timeGroupMap = new Map();
+	shippingTimes.forEach( ( { countryCode, time } ) => {
+		const group = timeGroupMap.get( time ) || { countryCodes: [], time };
+		group.countryCodes.push( countryCode );
+		timeGroupMap.set( time, group );
+	} );
+
+	return Array.from( timeGroupMap.values() );
+};
```

Comment:
❓
Apart from the property name `countryCodes`, the `getShippingTimesGroups` function is completely the same as the two existing functions:

- [getCountriesTimeArray](https://github.com/woocommerce/google-listings-and-ads/blob/be3ad280d25d7177e725f7a2556c1c4281f7f0a6/js/src/setup-mc/setup-stepper/setup-free-listings/shipping-time/shipping-time-setup/getCountriesTimeArray.js)
- [getCountriesTimeArray](https://github.com/woocommerce/google-listings-and-ads/blob/be3ad280d25d7177e725f7a2556c1c4281f7f0a6/js/src/components/free-listings/configure-product-listings/shipping-time/shipping-time-setup/getCountriesTimeArray.js)

Would you think they could be reused? Or even further reduced to a single one?

Reply:
That's a good idea actually, to reduced to a single one. I forgot about the `getCountriesTimeArray` functions, and I created the `getShippingTimesGroups` based on the deleted `saveShippingData` (https://github.com/woocommerce/google-listings-and-ads/pull/1420/files#diff-5297a79a23cac9dd2adeb1b6289484c6a7ebe09f096a488d19fd7ac68213fd1dL61-L77). 😂 

However, I'm not very keen in doing it in this PR. If I were to do it now, it means the Setup MC area would be impacted as well, and make this PR more complicated. Also, `getShippingTimesGroups` has a slightly different structure in its returned data `{ countryCodes, time }`, while `getCountriesTimeArray` has `{ countries, time }`. 

I'm thinking to have a code refactor on shipping time section to make it similar with shipping rate section (hopefully soon). This means that eventually we will have one copy of `getCountriesTimeArray` instead of two copies. I can think of how to consolidate `getShippingTimesGroups` and `getCountriesTimeArray` at that time (say, changing `getCountriesTimeArray` to return `{ countryCodes, time }`, since upsert shipping time works with that structure), and come out with tests for it.

I'll come out with a follow-up issue on this.

Reply:
I've created issue https://github.com/woocommerce/google-listings-and-ads/issues/1426 for this.